File: leetcode/problems/language-specific/go/387.first-unique-character-in-a-string.go
/*
 * @lc app=leetcode id=387 lang=golang
 *
 * [387] First Unique Character in a String
 */

// @lc code=start
func firstUniqChar(s string) int {
	if len(s) == 1 {
		return 0
	}

	mappings := [26]int{}
	for _, v := range s {
		mappings[int(v-'a')]++
	}

	for i, v := range s {
		if mappings[int(v-'a')] == 1 {
			return i
		}
	}

	return -1
}

// @lc code=end

File: leetcode/problems/language-specific/python/387.first-unique-character-in-a-string.py
#
# @lc app=leetcode id=387 lang=python3
#
# [387] First Unique Character in a String
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(1) Space
# * n -> The length of s string

class Solution:
    def firstUniqChar(self, s: str) -> int:
        # * Contains only 26 lowercase letters which makes it a constant space
        s_counter = collections.Counter(s)
        for idx, c in enumerate(s):
            if s_counter[c] == 1:
                return idx

        return -1

# @lc code=end
